File: static/admin-bookings.js
"use strict";
// Session token lives only in this tab's memory; never in persistent browser storage.
let token = "",
  selected = null,
  cursor = null,
  busy = false;
const el = (id) => document.getElementById(id);
const labels = {
  pending: "ফল আসার অপেক্ষায়",
  outcome_unknown: "ফল নিশ্চিত নয়",
  manually_resolved: "সমাধান নথিভুক্ত",
  held: "Hold আছে",
  not_created: "Booking তৈরি হয়নি",
  issued: "Ticket issued",
  cancelled: "Cancelled",
};
const errors = {
  MANUAL_RECONCILIATION_REQUIRED:
    "Supplier reference পাওয়া যায়নি। Supplier portal বা support থেকে যাচাই করুন।",
  SUPPLIER_RECONCILIATION_FAILED:
    "Supplier থেকে নিশ্চিত status পাওয়া যায়নি। Portal বা support থেকে যাচাই করুন।",
  SUPPLIER_READ_FAILED: "Supplier-এর সঙ্গে সংযোগ হয়নি। পরে আবার চেষ্টা করুন।",
  SUPPLIER_TIMEOUT: "Supplier উত্তর দিতে সময় নিচ্ছে। পরে status চেক করুন।",
  SUPPLIER_SERVICING_DISABLED:
    "এই supplier-এর status check বন্ধ আছে। দায়িত্বপ্রাপ্ত ব্যক্তির সঙ্গে যোগাযোগ করুন।",
  STALE_BOOKING_VIEW:
    "এই booking আপডেট হয়েছে। আবার খুলে তথ্য দেখে সিদ্ধান্ত দিন।",
  ALREADY_RESOLVED: "এই booking আগেই resolve করা হয়েছে। আবার খুলে ফল দেখুন।",
  BOOKING_NOT_RESOLVABLE: "Request এখনও চলতে পারে। কিছুক্ষণ পরে আবার দেখুন।",
  SUPPLIER_EVIDENCE_REQUIRED:
    "Supplier confirmation, কারণ, প্রমাণ ও reference পূরণ করুন।",
};
function message(text) {
  el("message").textContent = text;
}
function signedOut() {
  token = "";
  selected = null;
  cursor = null;
  el("workspace").hidden = true;
  el("login").hidden = false;
  el("logout").hidden = true;
  el("bookings").replaceChildren();
  el("detail").replaceChildren();
  el("environment").textContent = "Admin";
}
async function api(path, method = "GET", data) {
  const r = await fetch(path, {
    method,
    headers: {
      "Content-Type": "application/json",
      ...(token ? { Authorization: "Bearer " + token } : {}),
    },
    ...(data ? { body: JSON.stringify(data) } : {}),
  });
  let body = {};
  try {
    body = await r.json();
  } catch {}
  if (!r.ok) {
    if (r.status === 401) {
      signedOut();
      throw Error("লগইন করুন। Session শেষ হয়ে থাকতে পারে বা তথ্য সঠিক নয়।");
    }
    throw Error(
      errors[body.error] ||
        "কাজটি সম্পন্ন হয়নি। তথ্য যাচাই করে আবার চেষ্টা করুন।",
    );
  }
  return body;
}
async function action(fn) {
  if (busy) return;
  busy = true;
  document.querySelectorAll("button").forEach((b) => (b.disabled = true));
  try {
    message("");
    await fn();
  } catch (e) {
    message(e.message);
  } finally {
    busy = false;
    document.querySelectorAll("button").forEach((b) => (b.disabled = false));
  }
}
function node(tag, text, cls) {
  const n = document.createElement(tag);
  if (text !== undefined) n.textContent = text;
  if (cls) n.className = cls;
  return n;
}
async function list(append = false) {
  const resolved = el("view").value === "resolved";
  const body = await api(
    "/admin/bookings?resolved=" +
      resolved +
      (append && cursor ? "&before=" + encodeURIComponent(cursor) : ""),
  );
  el("queueTitle").textContent = resolved
    ? "সমাধান নথিভুক্ত"
    : "যাচাইয়ের অপেক্ষায়";
  el("queueDescription").textContent = resolved
    ? "নথিভুক্ত সিদ্ধান্ত ও তার প্রমাণ দেখতে একটি booking নির্বাচন করুন।"
    : "Supplier থেকে নিশ্চিত ফল না পাওয়া bookingগুলো এখানে দেখুন।";
  if (!append) el("bookings").replaceChildren();
  el("environment").textContent = body.environment;
  cursor = body.nextCursor;
  el("more").hidden = !cursor;
  for (const b of body.items) {
    const button = node(
      "button",
      (b.publicRef || "Reference pending") + " · " + b.clientName + " · " + (b.pnr || "PNR পাওয়া যায়নি"),
      "booking",
    );
    button.append(
      node("span", labels[b.state] + " · " + b.supplier),
      node("span", new Date(b.createdAt).toLocaleString("bn-BD")),
    );
    button.onclick = () => action(() => open(b.id));
    el("bookings").append(button);
  }
  if (!el("bookings").children.length)
    el("bookings").append(node("p", "এই তালিকায় কোনো booking নেই।", "muted"));
}
async function open(id) {
  selected = await api("/admin/bookings/" + id);
  render();
}
function render() {
  const b = selected,
    d = el("detail");
  d.replaceChildren(node("h2", "Booking বিস্তারিত"));
  const facts = node("dl");
  for (const [name, value] of [
    ["Agency", b.clientName],
    ["Order Ref", b.publicRef || "PNR পাওয়ার অপেক্ষায়"],
    ["Booking ID", b.id],
    ["Supplier", b.supplier],
    ["PNR", b.pnr || "পাওয়া যায়নি"],
    ["আমাদের রেকর্ড", labels[b.state] || b.state],
    ["Supplier status", b.supplierStatus || "নিশ্চিত নয়"],
    ["Ticket deadline", b.deadline || "নিশ্চিত নয়"],
    [
      "শেষ status check",
      b.lastCheckedAt
        ? new Date(b.lastCheckedAt).toLocaleString("bn-BD")
        : "এখনও হয়নি",
    ],
  ]) {
    facts.append(node("dt", name), node("dd", value));
  }
  d.append(facts);
  const recheck = node("button", "Status আবার চেক করুন", "secondary");
  recheck.onclick = () =>
    action(async () => {
      try {
        await api("/admin/bookings/" + b.id + "/recheck", "POST", {});
        message(
          "Supplier status পাওয়া গেছে। তথ্য দেখে প্রয়োজন হলে সিদ্ধান্ত দিন।",
        );
      } finally {
        await open(b.id);
      }
    });
  d.append(recheck);
  if (b.resolutionHistory && b.resolutionHistory.length) {
    const history = node("details"),
      summary = node(
        "summary",
        "সিদ্ধান্তের ইতিহাস (" + b.resolutionHistory.length + ")",
      );
    history.append(summary);
    for (const r of b.resolutionHistory) {
      const entry = node("dl");
      for (const [name, value] of [
        ["ফল", labels[r.outcome]],
        ["Admin", r.administratorName],
        ["সময়", new Date(r.createdAt).toLocaleString("bn-BD")],
        ["কারণ", r.reason],
        ["প্রমাণ", r.evidence],
        ["Reference", r.supplierCaseRef],
      ])
        entry.append(node("dt", name), node("dd", value));
      history.append(entry);
    }
    d.append(history);
  }
  if (b.lateOutcome)
    d.append(
      node(
        "p",
        "আগের সিদ্ধান্তের পরে supplier response এসেছে। নতুন তথ্য যাচাই করে আবার সিদ্ধান্ত দিন।",
        "notice",
      ),
    );
  if (b.resolution && b.state === "manually_resolved") {
    d.append(node("h2", "নথিভুক্ত সিদ্ধান্ত"));
    const res = node("dl");
    for (const [k, v] of [
      ["ফল", labels[b.resolution.outcome]],
      ["Admin", b.resolution.administratorName],
      ["কারণ", b.resolution.reason],
      ["প্রমাণ", b.resolution.evidence],
      ["Support reference", b.resolution.supplierCaseRef],
      ["সময়", new Date(b.resolution.createdAt).toLocaleString("bn-BD")],
    ])
      res.append(node("dt", k), node("dd", v));
    d.append(res);
    return;
  }
  if (!b.canResolve) {
    d.append(
      node(
        "p",
        "Request এখনও চলতে পারে। পাঁচ মিনিটের কম পুরোনো pending booking resolve করা যায় না। পরে তালিকা আপডেট করুন।",
        "notice",
      ),
    );
    return;
  }
  d.append(
    node(
      "p",
      "Supplier portal বা support থেকে নিশ্চিত হওয়ার পর সিদ্ধান্ত দিন। এই কাজ নতুন Hold, Issue বা Cancel করে না।",
      "notice",
    ),
  );
  const form = node("form");
  const choose = node("select");
  choose.id = "outcome";
  for (const [v, t] of [
    ["", "ফল নির্বাচন করুন"],
    ["held", "Supplier নিশ্চিত করেছে: Hold আছে"],
    ["not_created", "Supplier নিশ্চিত করেছে: Booking তৈরি হয়নি"],
    ["issued", "Supplier নিশ্চিত করেছে: Ticket issued"],
    ["cancelled", "Supplier নিশ্চিত করেছে: Cancelled"],
  ]) {
    const o = node("option", t);
    o.value = v;
    choose.append(o);
  }
  choose.required = true;
  const l = node("label", "যাচাইয়ের ফল");
  l.htmlFor = "outcome";
  form.append(l, choose);
  for (const [id, label, min, max, tag] of [
    ["reason", "কেন এই সিদ্ধান্ত নিচ্ছেন?", 10, 2000, "textarea"],
    [
      "evidence",
      "কীভাবে যাচাই করেছেন? Supplier-এর নিশ্চিত তথ্য লিখুন।",
      10,
      4000,
      "textarea",
    ],
    [
      "supplierCaseRef",
      "Supplier support case / confirmation reference",
      3,
      200,
      "input",
    ],
  ]) {
    const labelEl = node("label", label);
    labelEl.htmlFor = id;
    const inp = node(tag);
    inp.id = id;
    inp.required = true;
    inp.minLength = min;
    inp.maxLength = max;
    form.append(labelEl, inp);
  }
  form.append(
    node(
      "p",
      "Passport, password বা পূর্ণ passenger details এখানে লিখবেন না।",
      "muted",
    ),
  );
  const check = node("label", undefined, "check"),
    box = node("input");
  box.type = "checkbox";
  box.required = true;
  check.append(
    box,
    node("span", "আমি supplier-এর সঙ্গে যাচাই করেছি এবং উপরের ফল নিশ্চিত।"),
  );
  form.append(check);
  const buttons = node("div", undefined, "actions"),
    save = node("button", "সিদ্ধান্ত নথিভুক্ত করুন");
  buttons.append(save);
  form.append(buttons);
  form.onsubmit = (e) => {
    e.preventDefault();
    action(async () => {
      await api("/admin/bookings/" + b.id + "/resolve", "POST", {
        outcome: choose.value,
        reason: el("reason").value,
        evidence: el("evidence").value,
        supplierCaseRef: el("supplierCaseRef").value,
        expectedUpdatedAt: b.updatedAt,
        confirmedWithSupplier: box.checked,
      });
      await list();
      await open(b.id);
      message("সিদ্ধান্ত ও audit record সংরক্ষিত হয়েছে।");
    });
  };
  d.append(form);
}
el("loginForm").onsubmit = (e) => {
  e.preventDefault();
  action(async () => {
    const body = await api("/admin/login", "POST", {
      username: el("username").value,
      password: el("password").value,
    });
    token = body.access_token;
    el("password").value = "";
    el("login").hidden = true;
    el("workspace").hidden = false;
    el("logout").hidden = false;
    await list();
  });
};
el("refresh").onclick = () => action(() => list());
el("more").onclick = () => action(() => list(true));
el("logout").onclick = () =>
  action(async () => {
    try {
      await api("/admin/logout", "POST", {});
    } finally {
      signedOut();
    }
  });

el("view").onchange = () =>
  action(async () => {
    cursor = null;
    await list();
  });

File: src/admin_bookings.rs
//! Human-admin reconciliation. No supplier mutation is available here.
use crate::{
    AppState,
    auth::{Admin, ApiError},
};
use axum::{
    Json, Router,
    extract::{Path, Query, State},
    http::{HeaderMap, StatusCode},
    response::Html,
    routing::{get, post},
};
use serde::Deserialize;
use serde_json::{Value, json};
use utoipa::{OpenApi, ToSchema};
use uuid::Uuid;

fn conflict(code: &'static str) -> ApiError {
    ApiError(StatusCode::CONFLICT, code)
}
#[derive(Deserialize, Default)]
pub struct Page {
    #[serde(default)]
    before: Option<Uuid>,
    #[serde(default)]
    resolved: bool,
    #[serde(default)]
    cancellations: bool,
    #[serde(default)]
    cancellation_pending: bool,
}
#[utoipa::path(get,path="/admin/bookings",operation_id="admin_booking_list",tag="Booking reconciliation",security(("admin_session"=[])),params(("before"=Option<String>,Query,description="Cursor from nextCursor"),("resolved"=Option<bool>,Query,description="Show manually resolved records instead of unresolved queue"),("cancellations"=Option<bool>,Query,description="Show cancellation records with their independent outcome")),responses((status=200,body=Object)))]
async fn list(
    _admin: Admin,
    State(state): State<AppState>,
    Query(page): Query<Page>,
) -> Result<Json<Value>, ApiError> {
    let rows: Vec<(Value,)> = sqlx::query_as("SELECT jsonb_build_object('id',b.id,'publicRef',b.public_ref,'clientName',c.name,'supplier',b.supplier_id,'state',b.state,'cancellationState',(SELECT state FROM flight_cancellations x WHERE x.booking_id=b.id),'pnr',b.pnr,'createdAt',b.created_at) FROM flight_bookings b JOIN api_clients c ON c.id=b.client_id WHERE (CASE WHEN $3 THEN EXISTS(SELECT 1 FROM flight_cancellations x WHERE x.booking_id=b.id AND (NOT $4 OR x.state IN ('pending','outcome_unknown'))) WHEN $2 THEN b.state='manually_resolved' ELSE b.state IN ('pending','outcome_unknown') END) AND ($1::uuid IS NULL OR b.id < $1) ORDER BY b.id DESC LIMIT 51")
        .bind(page.before).bind(page.resolved).bind(page.cancellations).bind(page.cancellation_pending).fetch_all(&state.pool).await?;
    let mut items: Vec<Value> = rows.into_iter().map(|(v,)| v).collect();
    let next = if items.len() > 50 {
        items.truncate(50);
        items.last().map(|v| v["id"].clone())
    } else {
        None
    };
    Ok(Json(
        json!({"items":items,"nextCursor":next,"environment":state.environment}),
    ))
}
#[utoipa::path(get,path="/admin/bookings/{id}",operation_id="admin_booking_detail",tag="Booking reconciliation",security(("admin_session"=[])),params(("id"=String,Path)),responses((status=200,body=Object),(status=404)))]
async fn detail(
    _admin: Admin,
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<Json<Value>, ApiError> {
    let row: Option<(Value,)> = sqlx::query_as("SELECT jsonb_build_object('id',b.id,'publicRef',b.public_ref,'clientName',c.name,'supplier',b.supplier_id,'state',b.state,'pnr',b.pnr,'createdAt',b.created_at,'updatedAt',b.updated_at,'canResolve',(b.state='outcome_unknown' OR (b.state='pending' AND b.created_at < clock_timestamp()-INTERVAL '5 minutes')),'supplierStatus',CASE WHEN b.last_reconciliation_verified THEN b.last_reconciliation#>'{item1,status}' ELSE NULL END,'deadline',b.ticketing_time_limit,'lastCheckedAt',b.reconciled_at,'lateOutcome',b.error_code='LATE_BOOKING_OUTCOME','resolution',CASE WHEN r.booking_id IS NULL THEN NULL ELSE jsonb_build_object('outcome',r.outcome,'reason',r.reason,'evidence',r.evidence,'supplierCaseRef',r.supplier_case_ref,'administratorId',r.administrator_id,'administratorName',r.administrator_name,'createdAt',r.created_at) END,'resolutionHistory',(SELECT COALESCE(jsonb_agg(jsonb_build_object('outcome',h.outcome,'reason',h.reason,'evidence',h.evidence,'supplierCaseRef',h.supplier_case_ref,'administratorName',a.username,'createdAt',h.created_at) ORDER BY h.id DESC),'[]'::jsonb) FROM booking_resolutions h JOIN administrators a ON a.id=h.administrator_id WHERE h.booking_id=b.id)) FROM flight_bookings b JOIN api_clients c ON c.id=b.client_id LEFT JOIN LATERAL (SELECT br.*,a.username AS administrator_name FROM booking_resolutions br JOIN administrators a ON a.id=br.administrator_id WHERE br.booking_id=b.id ORDER BY br.id DESC LIMIT 1) r ON true WHERE b.id=$1")
        .bind(id).fetch_optional(&state.pool).await?;
    let mut detail = row.ok_or(ApiError(StatusCode::NOT_FOUND, "NOT_FOUND"))?.0;
    let cancellation:Option<(Value,)> = sqlx::query_as("SELECT jsonb_build_object('id',c.id,'state',c.state,'createdAt',c.created_at,'updatedAt',c.updated_at,'requiresReview',c.state<>'cancelled','evidence',(SELECT COALESCE(jsonb_agg(jsonb_build_object('verifiedCancelled',r.verified_cancelled,'responseAvailable',r.response IS NOT NULL,'checkedAt',r.created_at) ORDER BY r.id DESC),'[]'::jsonb) FROM (SELECT * FROM flight_cancellation_reconciliations WHERE cancellation_id=c.id ORDER BY id DESC LIMIT 50) r)) FROM flight_cancellations c WHERE c.booking_id=$1").bind(id).fetch_optional(&state.pool).await?;
    detail["cancellation"] = cancellation.map(|(v,)| v).unwrap_or(Value::Null);
    Ok(Json(detail))
}
#[utoipa::path(post,path="/admin/bookings/{id}/recheck",operation_id="admin_booking_recheck",tag="Booking reconciliation",security(("admin_session"=[])),params(("id"=String,Path)),responses((status=200,body=Object),(status=409,description="Supplier references missing; use supplier support"),(status=502),(status=504)))]
async fn recheck(
    admin: Admin,
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<(HeaderMap, Json<Value>), ApiError> {
    let (client,): (Uuid,) = sqlx::query_as("SELECT client_id FROM flight_bookings WHERE id=$1")
        .bind(id)
        .fetch_optional(&state.pool)
        .await?
        .ok_or(ApiError(StatusCode::NOT_FOUND, "NOT_FOUND"))?;
    crate::booking::reconcile_owned(&state, client, id, "admin", admin.id).await
}
#[derive(Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum Outcome {
    Held,
    NotCreated,
    Issued,
    Cancelled,
}
impl Outcome {
    fn value(&self) -> &'static str {
        match self {
            Self::Held => "held",
            Self::NotCreated => "not_created",
            Self::Issued => "issued",
            Self::Cancelled => "cancelled",
        }
    }
}
#[derive(Deserialize, ToSchema)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Resolution {
    outcome: Outcome,
    reason: String,
    evidence: String,
    supplier_case_ref: String,
    expected_updated_at: String,
    confirmed_with_supplier: bool,
}
fn valid_text(s: &str, min: usize, max: usize) -> bool {
    let n = s.trim().chars().count();
    n >= min && n <= max && !s.chars().any(|c| c.is_control() && c != '\n')
}
#[utoipa::path(post,path="/admin/bookings/{id}/resolve",operation_id="admin_booking_resolve",tag="Booking reconciliation",security(("admin_session"=[])),params(("id"=String,Path)),request_body=Resolution,responses((status=200,body=Object),(status=409,description="Stale view, already resolved, or request still in flight"),(status=422,description="Supplier confirmation and evidence required")))]
async fn resolve(
    admin: Admin,
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
    Json(input): Json<Resolution>,
) -> Result<Json<Value>, ApiError> {
    if !input.confirmed_with_supplier
        || !valid_text(&input.reason, 10, 2000)
        || !valid_text(&input.evidence, 10, 4000)
        || !valid_text(&input.supplier_case_ref, 3, 200)
    {
        return Err(ApiError(
            StatusCode::UNPROCESSABLE_ENTITY,
            "SUPPLIER_EVIDENCE_REQUIRED",
        ));
    }
    let expected = chrono::DateTime::parse_from_rfc3339(&input.expected_updated_at)
        .map_err(|_| ApiError(StatusCode::UNPROCESSABLE_ENTITY, "INVALID_VERSION"))?
        .with_timezone(&chrono::Utc);
    let mut tx = state.pool.begin().await?;
    let row: Option<(String,chrono::DateTime<chrono::Utc>,bool)> = sqlx::query_as("SELECT state,updated_at,created_at < clock_timestamp()-INTERVAL '5 minutes' FROM flight_bookings WHERE id=$1 FOR UPDATE")
        .bind(id).fetch_optional(&mut *tx).await?;
    let (old, version, aged) = row.ok_or(ApiError(StatusCode::NOT_FOUND, "NOT_FOUND"))?;
    if old == "manually_resolved" {
        return Err(conflict("ALREADY_RESOLVED"));
    }
    if version != expected {
        return Err(conflict("STALE_BOOKING_VIEW"));
    }
    if old != "outcome_unknown" && !(old == "pending" && aged) {
        return Err(conflict("BOOKING_NOT_RESOLVABLE"));
    }
    let outcome = input.outcome.value();
    sqlx::query("INSERT INTO booking_resolutions(booking_id,administrator_id,outcome,reason,evidence,supplier_case_ref,previous_state) VALUES($1,$2,$3,$4,$5,$6,$7)")
        .bind(id).bind(admin.id).bind(outcome).bind(input.reason.trim()).bind(input.evidence.trim()).bind(input.supplier_case_ref.trim()).bind(&old).execute(&mut *tx).await?;
    // This is an explicit manual outcome, never a fabricated verified hold payload.
    let public = json!({"bookingId":id,"state":"manually_resolved","requiresReconciliation":false,"manualOutcome":outcome});
    sqlx::query("UPDATE flight_bookings SET state='manually_resolved',public_response=$2,error_code=NULL,updated_at=clock_timestamp() WHERE id=$1").bind(id).bind(&public).execute(&mut *tx).await?;
    sqlx::query("INSERT INTO audit_events(actor_kind,actor_id,action,resource_kind,resource_id,metadata) VALUES('admin',$1,'booking.manual_resolution','booking',$2,$3)")
        .bind(admin.id.to_string()).bind(id.to_string()).bind(json!({"previousState":old,"outcome":outcome})).execute(&mut *tx).await?;
    tx.commit().await?;
    Ok(Json(public))
}
async fn panel() -> (HeaderMap, Html<&'static str>) {
    let mut h = HeaderMap::new();
    h.insert("content-security-policy", "default-src 'none'; script-src 'self'; style-src 'unsafe-inline'; connect-src 'self'; img-src 'self'; base-uri 'none'; form-action 'self'; frame-ancestors 'none'".parse().unwrap());
    h.insert("x-content-type-options", "nosniff".parse().unwrap());
    h.insert("referrer-policy", "no-referrer".parse().unwrap());
    (h, Html(include_str!("../static/admin-bookings.html")))
}
async fn script() -> (HeaderMap, &'static str) {
    let mut h = HeaderMap::new();
    h.insert(
        "content-type",
        "text/javascript; charset=utf-8".parse().unwrap(),
    );
    (h, include_str!("../static/admin-bookings.js"))
}
#[derive(OpenApi)]
#[openapi(
    paths(list, detail, recheck, resolve),
    components(schemas(Resolution, Outcome))
)]
pub struct AdminBookingDoc;
pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/admin/reconciliation", get(panel))
        .route("/admin/reconciliation.js", get(script))
        .route("/admin/bookings", get(list))
        .route("/admin/bookings/{id}", get(detail))
        .route("/admin/bookings/{id}/recheck", post(recheck))
        .route("/admin/bookings/{id}/resolve", post(resolve))
}
